<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />



<meta name="date" content="2025-02-10" />

<title>Reproducible scripting, AlphaSimR, Homework</title>

<script src="site_libs/header-attrs-2.28/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.13.2/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/main/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>



<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">Lab_3_7170_workflow</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Reproducible scripting, AlphaSimR,
Homework</h1>
<h4 class="author">Dinesh Ghimire</h4>
<address class="author_afil">
Cornell<br><h4 class="date">February 10, 2025</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span>
workflowr <span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span
class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> </a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2025-02-10
</p>
<p>
<strong>Checks:</strong> <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 6
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> 1
</p>
<p>
<strong>Knit directory:</strong> <code>Lab_3_7170_workflow/</code> <span
class="glyphicon glyphicon-question-sign" aria-hidden="true"
title="This is the local directory in which the code in this file was executed.">
</span>
</p>
<p>
This reproducible <a href="https://rmarkdown.rstudio.com">R Markdown</a>
analysis was created with <a
  href="https://github.com/workflowr/workflowr">workflowr</a> (version
1.7.1). The <em>Checks</em> tab describes the reproducibility checks
that were applied when the results were created. The <em>Past
versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguncommittedchanges">
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> <strong>R Markdown file:</strong> uncommitted
changes </a>
</p>
</div>
<div id="strongRMarkdownfilestronguncommittedchanges"
class="panel-collapse collapse">
<div class="panel-body">
<p>The R Markdown is untracked by Git. To know which version of the R
Markdown file created these results, you’ll want to first commit it to
the Git repo. If you’re still working on the analysis, you can ignore
this warning. When you’re finished, you can run
<code>wflow_publish</code> to commit the R Markdown file and build the
HTML.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the
global environment can affect the analysis in your R Markdown file in
unknown ways. For reproduciblity it’s best to always run the code in an
empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20250204code">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Seed:</strong>
<code>set.seed(20250204)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20250204code"
class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20250204)</code> was run prior to running
the code in the R Markdown file. Setting a seed ensures that any results
that rely on randomness, e.g. subsampling or permutations, are
reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Session information:</strong>
recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package
versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be
confident that you successfully produced the results during this
run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr
project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongfc7dbba">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Repository version:</strong> fc7dbba
</a>
</p>
</div>
<div id="strongRepositoryversionstrongfc7dbba"
class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development
and connecting the code version to the results is critical for
reproducibility.
</p>
<p>
The results in this page were generated with repository version fc7dbba.
See the <em>Past versions</em> tab to see a history of the changes made
to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for
the analysis have been committed to Git prior to generating the results
(you can use <code>wflow_publish</code> or
<code>wflow_git_commit</code>). workflowr only checks the R Markdown
file, but you know if there are other scripts or data files that it
depends on. Below is the status of the Git repository when the results
were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .DS_Store
    Ignored:    .Rproj.user/

Untracked files:
    Untracked:  analysis/Lab3_ProgrammingAlphaSimR.Rmd
    Untracked:  code/addToREADME.R
    Untracked:  output/QTLgenotypes.csv

Unstaged changes:
    Modified:   README.md

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not
included in this status report because it is ok for generated content to
have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">
<p>
There are no past versions. Publish this analysis with
<code>wflow_publish()</code> to start tracking its development.
</p>
<hr>
</div>
</div>
</div>
<style>
pre {
  overflow-x: auto;
}
pre code {
  word-wrap: normal;
  white-space: pre;
}
</style>
<div id="preliminaries" class="section level1">
<h1>Preliminaries</h1>
<div id="rmarkdown" class="section level3">
<h3>Rmarkdown</h3>
<p>This lab is written in Rmarkdown. I find it challenging to remember
the syntax, but here is a <a
href="https://rstudio.github.io/cheatsheets/html/rmarkdown.html">cheatsheet</a></p>
</div>
<div id="learning-objectives" class="section level3">
<h3>Learning objectives</h3>
<ol style="list-style-type: decimal">
<li>A bit more on reproducible scripting<br />
Some other things that matter to me to ensure communicability and
reproducibility</li>
</ol>
<ol style="list-style-type: lower-roman">
<li>The main README.md file should contain information about the
conditions under which the script was run<br />
</li>
<li>I like the script to have a predictable strucure: load and document
the packages used, if there is a stochastic component, set the random
seed, set and document the parameter values that will affect the script
output<br />
</li>
<li>Use a consistent directory structure. Happily, such a consistent
structure is enforced by <code>workflowr</code><br />
</li>
<li>Use the package <code>here</code> to make file locations relative to
the base project folder</li>
</ol>
<ol start="2" style="list-style-type: decimal">
<li><p>AlphaSimR<br />
<code>AlphaSimR</code> is a package to simulate breeding populations and
tasks. It is not completely intuitive nor completely well-documented. We
will want to use it ultimately to optimize breeding schemes. As a way to
introduce AlphaSimR, this Rmarkdown script contains code to illustrate
the “outbreak of variation” that occurs when a heterozygous individual
is self-fertilized. This phenomenon was important in supporting the
notion that quantitative traits were influenced by many loci. Here is an
<a
href="https://cran.r-project.org/web/packages/AlphaSimR/vignettes/intro.html">introduction</a>
showing code to simulate a very simple mass selection program. Here is a
<a
href="https://cran.r-project.org/web/packages/AlphaSimR/vignettes/traits.pdf">deep
dive</a> into how <code>AlphaSimR</code> simulates traits.</p></li>
<li><p>Prepare a Homework<br />
Your homework will be to write a script that illustrates the interesting
phenomenon of regression toward the mean between parents and progeny
that we discussed in class.</p></li>
</ol>
</div>
</div>
<div id="ordering-of-the-script" class="section level1">
<h1>Ordering of the script</h1>
<p>It’s good to have all scripts in the same order with respect to
standard tasks</p>
<div id="load-packages-first" class="section level3">
<h3>Load packages first</h3>
<p>If your script depends on external packages, load them at the
beginning. This shows users early on what the script dependencies
are.</p>
<pre class="r"><code>packages_used &lt;- c(&quot;AlphaSimR&quot;, &quot;tidyverse&quot;, &quot;workflowr&quot;, &quot;here&quot;)
ip &lt;- installed.packages()
all_packages_installed &lt;- TRUE
for (package in packages_used){
  if (!(package %in% ip[,&quot;Package&quot;])){
    print(paste(&quot;Please install package&quot;, package))
    all_packages_installed &lt;- FALSE
  }
}#END packages_used
if (!all_packages_installed) stop(&quot;Need to install packages&quot;)

library(tidyverse)</code></pre>
<pre><code>── Attaching core tidyverse packages ──────────────────────── tidyverse 2.0.0 ──
✔ dplyr     1.1.4     ✔ readr     2.1.5
✔ forcats   1.0.0     ✔ stringr   1.5.1
✔ ggplot2   3.5.1     ✔ tibble    3.2.1
✔ lubridate 1.9.3     ✔ tidyr     1.3.1
✔ purrr     1.0.2     
── Conflicts ────────────────────────────────────────── tidyverse_conflicts() ──
✖ dplyr::filter() masks stats::filter()
✖ dplyr::lag()    masks stats::lag()
ℹ Use the conflicted package (&lt;http://conflicted.r-lib.org/&gt;) to force all conflicts to become errors</code></pre>
<p>Notice the conflicts report from loading tidyverse. Two packages
(<code>dplyr</code> and <code>stats</code>) both have a function called
<code>filter</code>. Since dplyr was loaded <em>after</em> stats, if you
use the function <code>filter</code>, it will go to the dplyr version.
It is dangerous to rely on what order packages have been loaded to
determine which <code>filter</code> function you get. R syntax to
prevent ambiguity is to write either <code>dplyr::filter</code> or
<code>stats::filter</code>. Using that syntax will make your code more
reproducible.</p>
</div>
<div id="set-file-locations-relative-to-the-project-folder-using-here"
class="section level3">
<h3>Set file locations relative to the project folder using
<code>here</code></h3>
<pre class="r"><code>here::i_am(&quot;analysis/Lab3_ProgrammingAlphaSimR.Rmd&quot;)

# here::i_am is anchoring towards our project
# finding a base</code></pre>
<p>There are two functions I use in the <code>here</code> package:
<code>here::i_am</code> and <code>here::here</code>.
<code>here::i_am</code> helps <code>here</code> “understand” the folder
structure. You give it the name of the script that is currently running,
and the path to that script from the base of the overall analysis
folder. From that, <code>here</code> deduces where the base of the
folder is. After that, <code>here::here</code> locates files from that
folder base. If you use the <code>here</code> package, you can send
anyone a complicated zipped folder, and they should be able to run the
analysis out of the box, regardless of where they put the folder in
their own file system.</p>
</div>
<div id="document-packages-used" class="section level3">
<h3>Document packages used</h3>
<p>This chunk creates a “README.md” file that documents the packages and
versions used for future reference.</p>
</div>
<div
id="dg-notes-source-reads-whole-separate-file-to-accept-its-input-from-the-named-file-or-url-or-connection."
class="section level2">
<h2>DG notes: source reads whole separate file to accept its input from
the named file or URL or connection.</h2>
<pre class="r"><code>source(here::here(&quot;code/addToREADME.R&quot;))
addToREADME(paste0(&quot;## &quot;, rmarkdown::metadata$title), append=F)
addToREADME(c(date(), &quot;&quot;))

packages_info &lt;- ip[packages_used, c(&quot;Package&quot;, &quot;Version&quot;, &quot;Built&quot;)]
addToREADME(c(&quot;The packages used in this script are:&quot;, &quot;Package, Version, Built&quot;))
apply(packages_info, 1, function(vec) addToREADME(paste(vec, collapse=&quot; &quot;)))</code></pre>
<pre><code>              AlphaSimR               tidyverse               workflowr 
&quot;AlphaSimR 1.6.1 4.4.1&quot; &quot;tidyverse 2.0.0 4.4.0&quot; &quot;workflowr 1.7.1 4.4.0&quot; 
                   here 
     &quot;here 1.0.1 4.4.0&quot; </code></pre>
<pre class="r"><code>addToREADME(&quot;&quot;)</code></pre>
<p>Hyperlink the <a href="README.md">README file</a> to the report. That
makes it easy to find.</p>
<div id="set-random-seed" class="section level3">
<h3>Set random seed</h3>
<p>AlphaSimR generates many random numbers (e.g., to simulate Mendelian
random segregation). If you want the result of the analysis to come out
the same each time (there are pros and cons) you need to set the random
seed. Note that <code>workflowr</code> does this by default. If you are
not using that package, then be explicit.</p>
<pre class="r"><code>random_seed &lt;- 45678
set.seed(random_seed)
addToREADME(c(paste(&quot;The random seed is&quot;, random_seed), &quot;&quot;))</code></pre>
</div>
<div id="script-parameters" class="section level3">
<h3>Script parameters</h3>
<p>If the behavior of your script depends on parameters that you set,
initialize them early on.</p>
<pre class="r"><code>nFounders &lt;- 100 # starting population (eg. landraces or wild plants or animals)
nChr &lt;- 10 # Number of chromosomes
nSitesPerChr &lt;- 1000 # Number of segregating sites _per chromosome_
nQTLperChr &lt;- 10 # Vary this parameter to get oligo- versus poly- genic traits
nF1s &lt;- 200 # We are going to make F1s to test outbreak of variation
nF2s &lt;- 200 # We are going to make F2s to test outbreak of variation</code></pre>
</div>
<div id="parameters-to-readme" class="section level3">
<h3>Parameters to README</h3>
<p>It’s good to have all the information together in the README. Chunks
of code like this do not need to be included in the report. To exclude
them use the “include=FALSE” option in the chunk.</p>
</div>
</div>
</div>
<div id="simulating-some-classical-results" class="section level1">
<h1>Simulating some classical results</h1>
<p>This script uses <code>AlphaSimR</code> to simulate the “outbreak of
variation” that arises when you self-fertilize a hybrid.<br />
#DG notes: from AlphaSimR html file, there are generally 4 steps: 1.
Creating Founder Haplotypes 2.Setting Simulation Parameters 3. Modeling
the Breeding Program 4.Examining the Results</p>
<div id="alphasimr-populations" class="section level3">
<h3>AlphaSimR populations</h3>
<p>The basic object of <code>AlphaSimR</code> is the population. To make
founders, you first make founder haplotypes from a coalescent
simulation, then you define simulation parameters that will link their
genetic variation to phenotypic variation, then you make a first diploid
population from the founder haplotypes.</p>
<pre class="r"><code># 1. Creating Founder Haplotypes

# Create haplotypes for founder population of outbred individuals
# Note: default effective population size for runMacs is 100
founderHaps &lt;- AlphaSimR::runMacs(nInd=nFounders, nChr=nChr, 
                                  segSites=nSitesPerChr)
# founderHaps &lt;- AlphaSimR::quickHaplo(nInd=nFounders, nChr=nChr, 
#                                  segSites=nSitesPerChr)

# 2.Setting Simulation Parameters: New global simulation parameters from founder haplotypes
# i. first line initializes an object containing the simulation parameters
SP &lt;- AlphaSimR::SimParam$new(founderHaps)

# Additive trait architecture
# By default, the genetic variance will be 1
# ii. The next line of code defines the quantitative trait used for selection
SP$addTraitA(nQtlPerChr=nQTLperChr) 
### DG notes, A at the end is for additive effects only. for additive and dominance, there is AD and for additive and epistatis there is AE and so on.

# iii. # Create a new population of founders (now diploid created) by pairing random haplotypes to form diploid.
### in this steop, the founder haplotypes are paired together randomly to form diploid individuals, meaning each individual will have two haplotypes (from two parents)
founders &lt;- AlphaSimR::newPop(founderHaps, simParam=SP) # takes two random haplotype and makes diploir 
str(founders)</code></pre>
<pre><code>Formal class &#39;Pop&#39; [package &quot;AlphaSimR&quot;] with 18 slots
  ..@ id     : chr [1:100] &quot;1&quot; &quot;2&quot; &quot;3&quot; &quot;4&quot; ...
  ..@ iid    : int [1:100] 1 2 3 4 5 6 7 8 9 10 ...
  ..@ mother : chr [1:100] &quot;0&quot; &quot;0&quot; &quot;0&quot; &quot;0&quot; ...
  ..@ father : chr [1:100] &quot;0&quot; &quot;0&quot; &quot;0&quot; &quot;0&quot; ...
  ..@ sex    : chr [1:100] &quot;H&quot; &quot;H&quot; &quot;H&quot; &quot;H&quot; ...
  ..@ nTraits: int 1
  ..@ gv     : num [1:100, 1] -0.399 0.324 -1.687 -0.204 -0.411 ...
  .. ..- attr(*, &quot;dimnames&quot;)=List of 2
  .. .. ..$ : NULL
  .. .. ..$ : chr &quot;Trait1&quot;
  ..@ pheno  : num [1:100, 1] NA NA NA NA NA NA NA NA NA NA ...
  .. ..- attr(*, &quot;dimnames&quot;)=List of 2
  .. .. ..$ : NULL
  .. .. ..$ : chr &quot;Trait1&quot;
  ..@ ebv    : num[1:100, 0 ] 
  ..@ gxe    :List of 1
  .. ..$ : NULL
  ..@ fixEff : int [1:100] 1 1 1 1 1 1 1 1 1 1 ...
  ..@ misc   : list()
  ..@ miscPop: list()
  ..@ nInd   : int 100
  ..@ nChr   : int 10
  ..@ ploidy : int 2
  ..@ nLoci  : int [1:10] 1000 1000 1000 1000 1000 1000 1000 1000 1000 1000
  ..@ geno   :List of 10
  .. ..$ : raw [1:125, 1:2, 1:100] c5 07 ca 72 ...
  .. ..$ : raw [1:125, 1:2, 1:100] 60 d2 21 6c ...
  .. ..$ : raw [1:125, 1:2, 1:100] c0 51 18 3a ...
  .. ..$ : raw [1:125, 1:2, 1:100] 02 08 e8 34 ...
  .. ..$ : raw [1:125, 1:2, 1:100] 01 b0 dd 04 ...
  .. ..$ : raw [1:125, 1:2, 1:100] c7 00 fb c0 ...
  .. ..$ : raw [1:125, 1:2, 1:100] 51 c4 cc b9 ...
  .. ..$ : raw [1:125, 1:2, 1:100] 04 a8 18 90 ...
  .. ..$ : raw [1:125, 1:2, 1:100] 03 c2 4b 80 ...
  .. ..$ : raw [1:125, 1:2, 1:100] 90 89 bc 02 ...</code></pre>
<pre class="r"><code>#id&#39;s are character and iid are integers</code></pre>
</div>
<div id="population-information" class="section level3">
<h3>Population information</h3>
<p>The population has <code>id</code>s. The <code>@mother</code> and
<code>@father</code> ids are all zero because this population was made
from founder haplotypes, and so does not have diploid parents. The
genotypic values <code>gv</code> of the population are calculated for
the trait created using
<code>SP$addTraitA(nQtlPerChr=nQTLperChr)</code>. Given that there are
10 chromosomes and 10 QTL per chromosome, there are
<code>3^(nChr*nQTLperChr)</code> = 5.1537752^{47} different possible
genotypic values. The realized genotypic values are accessible with the
function <code>gv(founders)</code></p>
<p>From here, you can treat this population like a named vector using
the square braces extraction operator <code>[ ]</code>. Extract
individuals by their <code>@id</code> or just by their order in the
population using an integer index. For example, pick three random
individuals from a population and list their ids. Pick the one with the
first id in alphabetical order.</p>
<pre class="r"><code># here we have founders population. so we can access individuals from popn like we get numbers from a vector.
test &lt;- founders[c(2, 3, 5, 8, 13)]
testID &lt;- test@id # get id&#39;s
firstID &lt;- testID[order(testID)][1] #order those id&#39;s
alphaInd &lt;- test[firstID] # Extract individual with the first ID in order
print(testID)</code></pre>
<pre><code>[1] &quot;2&quot;  &quot;3&quot;  &quot;5&quot;  &quot;8&quot;  &quot;13&quot;</code></pre>
<pre class="r"><code>print(alphaInd)</code></pre>
<pre><code>An object of class &quot;Pop&quot; 
Ploidy: 2 
Individuals: 1 
Chromosomes: 10 
Loci: 10000 
Traits: 1 </code></pre>
<pre class="r"><code>print(alphaInd@id)</code></pre>
<pre><code>[1] &quot;13&quot;</code></pre>
</div>
<div id="outbreak-of-variation" class="section level3">
<h3>Outbreak of variation</h3>
<p>Emerson and East (1913) showed that if you crossed two inbreds, the
hybrid had similar variation to each inbred, but if you then selfed the
hybrid, the offspring varied more substantially. This code simulates
that result. First, self the founders to homozygosity. The function
<code>self</code> self-fertilizes individuals from the population. By
default, it creates one selfed individual per parent (controllable with
the parameter <code>nProgeny</code>), so this works nicely for
single-seed descent.</p>
<pre class="r"><code># Self-pollinate to for a few generations
nGenSelf &lt;- 3
inbredPop &lt;- founders
for (gen in 1:nGenSelf){
  inbredPop &lt;- AlphaSimR::self(inbredPop)
}

# does 3 generations of single seed descent</code></pre>
</div>
<div id="check-homozygosity" class="section level3">
<h3>Check homozygosity</h3>
<p>Just a sanity check that this has, in fact, created a population of
100 individuals that are appropriately homozygous. Loci are coded 0, 1,
2. So <code>qtl == 1</code> represents the case were a locus is
heterozygous. <code>sum(qtl == 1)</code> counts those cases.</p>
<pre class="r"><code>qtl &lt;- AlphaSimR::pullQtlGeno(inbredPop)
if (nrow(qtl) != nFounders) stop(&quot;The number of individuals is unexpected&quot;)
if (ncol(qtl) != nChr * nQTLperChr) stop(&quot;The number of QTL is unexpected&quot;)
fracHet &lt;- sum(qtl == 1) / (nFounders * nChr * nQTLperChr) # so here (nFounders * nChr * nQTLperChr) is total number of genotypes and when we set qtl==1, we are taking all heterozygous loci for fraction of Het.

cat(&quot;Expected fraction heterozygous&quot;, 1 / 2^nGenSelf, &quot;\n&quot;,
    &quot;Observed fraction heterozygous&quot;, fracHet, &quot;+/-&quot;, 
    round(2*sqrt(fracHet*(1-fracHet)/(nFounders*nChr*nQTLperChr)), 3), &quot;\n&quot;)</code></pre>
<pre><code>Expected fraction heterozygous 0.125 
 Observed fraction heterozygous 0.0357 +/- 0.004 </code></pre>
<pre class="r"><code># here with 1/ / 2^nGenSelf, we are assuming that after nGenSelf generation of selfing, the expected fraction of het loci will decrease exponentially. </code></pre>
<blockquote>
<p>What was wrong with my reasoning about the Expected fraction
heterozygous? not all loci are heterozygous to begin with (but we set
qtl ==1, so that means we only selected het loci)</p>
</blockquote>
</div>
<div id="simulate-outbreak-of-variation" class="section level3">
<h3>Simulate outbreak of variation</h3>
<p>We will<br />
1. pick a random pair of inbred individuals<br />
2. cross that pair<br />
3. find out the variation in genotypic value among the pair’s
progeny<br />
4. pick a random F1 progeny<br />
5. self-fertilize that F1<br />
6. find out the variation in genotypic value among the F1’s
progeny<br />
<em>It’s often useful to write very high-level outlines of what you want
your code to do before you write the code</em><br />
We will assume a trait that has a heritability of 0.5 in the base,
non-inbred population. An easy way to get that heritability is with
genetic and error variances of 1.</p>
</div>
<div id="pick-random-pair-to-make-f1-hybrid" class="section level3">
<h3>Pick random pair to make F1 hybrid</h3>
<pre class="r"><code>randomPair &lt;- inbredPop[sample(nFounders, 2)] # each time separate picked</code></pre>
</div>
<div id="cross-the-pair-to-make-population-of-f1s"
class="section level3">
<h3>Cross the pair to make population of F1s</h3>
<p>The crossPlan is a <strong>two-column matrix</strong> with as many
rows as the number of crosses you want to make, the first column is the
<code>@id</code> or the index of the seed parent, and likewise for the
pollen parent in the second column. We will make 200 F1s from this
random pair, so the matrix has 200 rows. You just want to cross
individual 1 with individual 2, so each row contains 1:2.</p>
<pre class="r"><code># here we first create a matrix with two columns and byrow = t (filled by rows, 1, 2, 1, 2..., maternal, paternal...). The data includes nF1&#39;s number of 1and 2 alternateively. 
# In this matrix, each row represents a cross between parent1(1) and parent 2(2).
crossPlan &lt;- matrix(rep(1:2, nF1s), ncol=2, byrow=T) # matrix(data = , nrow =, ncol =, byrow = , dimnames =)

#now makecross function perform actual crosses between pairs specified in the crossPlan above. so we have parent population (randomPair) and the crossing plan (crossPlan). thus as a result nf1&#39;s are created. 200 here
f1_pop &lt;- AlphaSimR::makeCross(randomPair, crossPlan) # to make crosses, use this function makeCross and for that we need popn and 
head(crossPlan)</code></pre>
<pre><code>     [,1] [,2]
[1,]    1    2
[2,]    1    2
[3,]    1    2
[4,]    1    2
[5,]    1    2
[6,]    1    2</code></pre>
<pre class="r"><code># here using crossPlan (which has 200 F1s in row with 1,2... meaning parent 1, parent2) and randomPair</code></pre>
</div>
<div id="find-out-the-phenotypic-variation-among-f1s"
class="section level3">
<h3>Find out the phenotypic variation among F1s</h3>
<p>When you first make a population, <code>AlphaSimR</code> does not
assume it has been phenotyped. You can phenotype it using the
<code>setPheno</code> function. Note that if you use
<code>setPheno</code> on the same population a second time, that will
overwrite the phenotypes from the first time. The genotypic variance can
be retrieved using the function <code>varG</code>. Really,
<code>varG</code> gives all we need to know, but of course that variance
is generally not observable in non-simulated reality.</p>
<pre class="r"><code>f1_pop &lt;- AlphaSimR::setPheno(f1_pop, varE=1) # error variance set as 1. initially in founders, genetic variance is 1. so Heritability is 1/(1+1) =0.5
cat(&quot;Genotypic variance among F1s&quot;, round(AlphaSimR::varG(f1_pop), 3), &quot;\n&quot;)</code></pre>
<pre><code>Genotypic variance among F1s 0.127 </code></pre>
<pre class="r"><code>cat(&quot;Phenotypic variance among F1s&quot;, round(AlphaSimR::varP(f1_pop), 3), &quot;\n&quot;)</code></pre>
<pre><code>Phenotypic variance among F1s 1.026 </code></pre>
<pre class="r"><code>hist(AlphaSimR::pheno(f1_pop), main=&quot;Histogram of F1 Phenotypes&quot;)</code></pre>
<p><img src="figure/Lab3_ProgrammingAlphaSimR.Rmd/Phenotype%20F1s-1.png" width="672" style="display: block; margin: auto;" />
# due to sampling error, phenotypic variance is not exactly equarl to
geno + error = 0.044 +1 + 1.044 (thereotically and if we run for
infinite times) but we got 0.941 due to sampling error. the values can
change…cz random</p>
</div>
<div id="pick-random-f1" class="section level3">
<h3>Pick random F1</h3>
<pre class="r"><code>randomF1 &lt;- f1_pop[sample(nFounders, 1)]</code></pre>
</div>
<div id="make-f2-and-observe-phenotypic-variance"
class="section level3">
<h3>Make F2 and observe phenotypic variance</h3>
<p>When you first make a population, <code>AlphaSimR</code> does not
assume it has been phenotyped. You can phenotype it using the
<code>setPheno</code> function. Note that if you use
<code>setPheno</code> on the same population a second time, that will
overwrite the phenotypes from the first time. The genotypic variance can
be retrieved using the function <code>varG</code>. Really,
<code>varG</code> gives all we need to know, but of course that variance
is generally not observable in non-simulated reality.</p>
<pre class="r"><code>f2_pop &lt;- AlphaSimR::self(randomF1, nProgeny=nF2s)
f2_pop &lt;- AlphaSimR::setPheno(f2_pop, varE=1)
cat(&quot;Genotypic variance among F2s&quot;, round(AlphaSimR::varG(f2_pop), 3), &quot;\n&quot;)</code></pre>
<pre><code>Genotypic variance among F2s 0.361 </code></pre>
<pre class="r"><code>cat(&quot;Phenotypic variance among F2s&quot;, round(AlphaSimR::varP(f2_pop), 3), &quot;\n&quot;)</code></pre>
<pre><code>Phenotypic variance among F2s 1.132 </code></pre>
<pre class="r"><code>hist(AlphaSimR::pheno(f2_pop), main=&quot;Histogram of F2 Phenotypes&quot;)</code></pre>
<p><img src="figure/Lab3_ProgrammingAlphaSimR.Rmd/Make%20F2-1.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code># par(mar = c(4, 4, 2, 1))  # to fit here</code></pre>
</div>
<div id="illustrate-herehere-by-writing-a-file" class="section level3">
<h3>Illustrate <code>here::here</code> by writing a file</h3>
<p>Just to practice again using the here package because it makes life
easy.</p>
<pre class="r"><code>utils::write.csv(qtl, here::here(&quot;output&quot;, &quot;QTLgenotypes.csv&quot;))</code></pre>
</div>
</div>
<div id="homework" class="section level1">
<h1>Homework</h1>
<p>Illustrate regression to the mean between parents and offspring using
AlphaSimR</p>
<p>Notes: My plan- - create parent population - makes crosses using
random mating and crossPlan. - generate F1’s - get phenotypes of parent
- get phenotypes of progeny - make a scatter plot - run linear model for
regression coefficient and interpret it</p>
<ol style="list-style-type: decimal">
<li>You know how to create a population – create a population of
parents</li>
</ol>
<pre class="r"><code>nFounders &lt;- 100
nChr &lt;- 20 # Number of chromosomes
nSitesPerChr &lt;- 1000 # Number of segregating sites _per chromosome_
nQTLperChr &lt;- 10 # we can vary this parameter to get oligo- versus poly- genic traits
nF1s &lt;- 200 # We are going to make F1s by crossing two parents to illustrate regression to the mean.</code></pre>
<pre class="r"><code># 1. Creating Founder Haplotypes

# Create haplotypes for founder population of outbred individuals
# Note: default effective population size for runMacs is 100
founderHaps &lt;- AlphaSimR::runMacs(nInd=nFounders, nChr=nChr, 
                                  segSites=nSitesPerChr)

# 2.Setting Simulation Parameters: New global simulation parameters from founder haplotypes
# i. first line initializes an object containing the simulation parameters
SP &lt;- AlphaSimR::SimParam$new(founderHaps)

# Additive trait architecture
# By default, the genetic variance will be 1
# ii. The next line of code defines the quantitative trait used for selection
SP$addTraitA(nQtlPerChr=nQTLperChr) 
### DG notes, A at the end is for additive effects only. for additive and dominance, there is AD and for additive and epistatis there is AE and so on.

# iii. # Create a new population of founders (now diploid created) by pairing random haplotypes to form diploid.
### in this step, the founder haplotypes are paired together randomly to form diploid individuals, meaning each individual will have two haplotypes (from two parents)
founders &lt;- AlphaSimR::newPop(founderHaps, simParam=SP) # takes two random haplotype and makes diploir 
str(founders)</code></pre>
<pre><code>Formal class &#39;Pop&#39; [package &quot;AlphaSimR&quot;] with 18 slots
  ..@ id     : chr [1:100] &quot;1&quot; &quot;2&quot; &quot;3&quot; &quot;4&quot; ...
  ..@ iid    : int [1:100] 1 2 3 4 5 6 7 8 9 10 ...
  ..@ mother : chr [1:100] &quot;0&quot; &quot;0&quot; &quot;0&quot; &quot;0&quot; ...
  ..@ father : chr [1:100] &quot;0&quot; &quot;0&quot; &quot;0&quot; &quot;0&quot; ...
  ..@ sex    : chr [1:100] &quot;H&quot; &quot;H&quot; &quot;H&quot; &quot;H&quot; ...
  ..@ nTraits: int 1
  ..@ gv     : num [1:100, 1] -0.331 1.234 -1.538 0.483 -0.606 ...
  .. ..- attr(*, &quot;dimnames&quot;)=List of 2
  .. .. ..$ : NULL
  .. .. ..$ : chr &quot;Trait1&quot;
  ..@ pheno  : num [1:100, 1] NA NA NA NA NA NA NA NA NA NA ...
  .. ..- attr(*, &quot;dimnames&quot;)=List of 2
  .. .. ..$ : NULL
  .. .. ..$ : chr &quot;Trait1&quot;
  ..@ ebv    : num[1:100, 0 ] 
  ..@ gxe    :List of 1
  .. ..$ : NULL
  ..@ fixEff : int [1:100] 1 1 1 1 1 1 1 1 1 1 ...
  ..@ misc   : list()
  ..@ miscPop: list()
  ..@ nInd   : int 100
  ..@ nChr   : int 20
  ..@ ploidy : int 2
  ..@ nLoci  : int [1:20] 1000 1000 1000 1000 1000 1000 1000 1000 1000 1000 ...
  ..@ geno   :List of 20
  .. ..$ : raw [1:125, 1:2, 1:100] 70 32 68 ca ...
  .. ..$ : raw [1:125, 1:2, 1:100] 1b 57 0e 02 ...
  .. ..$ : raw [1:125, 1:2, 1:100] 61 45 a5 20 ...
  .. ..$ : raw [1:125, 1:2, 1:100] cd 09 78 27 ...
  .. ..$ : raw [1:125, 1:2, 1:100] 3f a3 06 88 ...
  .. ..$ : raw [1:125, 1:2, 1:100] 40 35 22 00 ...
  .. ..$ : raw [1:125, 1:2, 1:100] 30 78 8c 22 ...
  .. ..$ : raw [1:125, 1:2, 1:100] 28 19 00 f2 ...
  .. ..$ : raw [1:125, 1:2, 1:100] 74 ac 21 54 ...
  .. ..$ : raw [1:125, 1:2, 1:100] 14 ea 1e 42 ...
  .. ..$ : raw [1:125, 1:2, 1:100] 70 a1 f1 df ...
  .. ..$ : raw [1:125, 1:2, 1:100] bc 85 16 13 ...
  .. ..$ : raw [1:125, 1:2, 1:100] 8e 08 25 38 ...
  .. ..$ : raw [1:125, 1:2, 1:100] 2a c0 82 b6 ...
  .. ..$ : raw [1:125, 1:2, 1:100] 88 82 e1 92 ...
  .. ..$ : raw [1:125, 1:2, 1:100] 63 80 c0 26 ...
  .. ..$ : raw [1:125, 1:2, 1:100] f3 76 c5 14 ...
  .. ..$ : raw [1:125, 1:2, 1:100] 05 88 03 a8 ...
  .. ..$ : raw [1:125, 1:2, 1:100] 05 0a 8e 50 ...
  .. ..$ : raw [1:125, 1:2, 1:100] 70 13 66 02 ...</code></pre>
<ol start="2" style="list-style-type: decimal">
<li>You know how to get phenotypes from that population</li>
</ol>
<pre class="r"><code># use setpheno package in AlphaSimR to get phenotypes and set varE to 1.
founders_pheno &lt;- AlphaSimR::setPheno(founders, varE=1)
pheno_founders &lt;- founders_pheno@pheno
pheno_founders</code></pre>
<pre><code>              Trait1
  [1,] -0.8588733918
  [2,]  0.2772265566
  [3,] -1.0727950621
  [4,]  1.2941694352
  [5,]  0.2399939079
  [6,] -0.4248865925
  [7,] -0.7674856130
  [8,]  1.6306937213
  [9,] -1.0560886211
 [10,]  0.5470052087
 [11,]  0.3305085889
 [12,]  0.1839449940
 [13,] -0.3936875908
 [14,]  0.7189815574
 [15,]  0.9385772823
 [16,] -0.3503885504
 [17,] -1.7086522250
 [18,]  1.4848110838
 [19,]  2.8249672965
 [20,] -0.8299388225
 [21,]  0.7688568742
 [22,] -0.9552328579
 [23,]  0.6604091197
 [24,]  0.2003897212
 [25,]  1.2175475857
 [26,] -1.1585508509
 [27,] -0.2573629784
 [28,] -0.7949148963
 [29,] -1.6340909561
 [30,] -1.1956780370
 [31,] -1.9957019726
 [32,]  0.2569412534
 [33,] -1.1698109147
 [34,]  2.0702953564
 [35,] -0.8108570660
 [36,]  0.1891954670
 [37,]  1.0558562265
 [38,]  1.8895102781
 [39,] -0.2396606646
 [40,] -1.0532264640
 [41,] -0.8569643481
 [42,]  0.6392518884
 [43,] -1.2507277980
 [44,] -1.2426629164
 [45,] -1.7549838626
 [46,]  0.4249442773
 [47,]  0.3009079754
 [48,]  0.7261296031
 [49,] -0.0245810048
 [50,] -0.6165908619
 [51,] -0.6197688004
 [52,] -2.7076488217
 [53,] -1.7242728815
 [54,] -1.3723200979
 [55,] -0.2377067454
 [56,]  1.8480615194
 [57,]  3.3439551806
 [58,] -0.2137220225
 [59,]  2.1762591805
 [60,]  0.6080678159
 [61,] -0.9534773248
 [62,]  2.6893945408
 [63,]  0.7265587143
 [64,] -0.0639638229
 [65,]  1.2928418820
 [66,] -0.0530108623
 [67,]  1.8638545769
 [68,] -0.1444587078
 [69,] -0.6241782483
 [70,] -0.0007862148
 [71,] -0.0773839826
 [72,] -1.6182005692
 [73,]  0.5037834604
 [74,]  3.3792864398
 [75,] -1.8241439782
 [76,]  0.5653903136
 [77,]  0.9757551536
 [78,]  2.4877238072
 [79,] -2.4209579050
 [80,]  0.7140382153
 [81,]  1.3181081944
 [82,] -2.0388940223
 [83,] -2.1680905670
 [84,]  0.9384906844
 [85,]  0.2896996057
 [86,]  3.4631463650
 [87,]  0.1210654474
 [88,] -1.2126618476
 [89,] -1.0324015119
 [90,] -2.8049187113
 [91,]  0.4298979855
 [92,]  1.3766630042
 [93,]  0.9755144785
 [94,] -0.6948051688
 [95,] -1.0420996049
 [96,]  1.2360131989
 [97,]  2.8558030397
 [98,]  0.7915352136
 [99,] -0.0134073986
[100,] -1.3839765778</code></pre>
<ol start="3" style="list-style-type: decimal">
<li>You want to randomly mate that population to get progeny that will
be (we assume) regressed to the mean. Create a <code>crossPlan</code>
matrix like I did to generate the F1s, except that each row should have
randomly-picked parents, rather than 1 and 2 like for the F1s. There is
also a command <code>AlphaSimR::randCross</code>. Check out its
documentation. If you use that command, you will have to find the
pedigree of the progeny using the <code>@mother</code> and
<code>@father</code> ids of the progeny population and match those up to
the parent population. If you make the crossPlan, then it gives you the
seed and pollen parent ids.</li>
</ol>
<pre class="r"><code># First lets create a crossing plan in a matrix form as above
# here i will create a matrix by randomly sampling 200(f1&#39;s needed) parents and since we have only 100 parents or founders, we set replace = TRUE. then we arrange that data into two columns and set it true by row.
crossPlan_2 &lt;- matrix(data = sample(x=nFounders, size = 200*2, replace = T), ncol = 2, byrow = T) # size 200*2 for 200 rows of 2 parents each ( thus total 400 parents selected from 100 founders meaning replacement true here)
head(crossPlan_2)</code></pre>
<pre><code>     [,1] [,2]
[1,]   69   91
[2,]   58   49
[3,]   98   40
[4,]   69   96
[5,]   92    6
[6,]   62   43</code></pre>
<pre class="r"><code># Second, since we already have randomly selected parents in crossPlan, we can cross them to create F1&#39;s.
f1_popn &lt;- AlphaSimR::makeCross(founders, crossPlan_2)
f1_popn</code></pre>
<pre><code>An object of class &quot;Pop&quot; 
Ploidy: 2 
Individuals: 200 
Chromosomes: 20 
Loci: 20000 
Traits: 1 </code></pre>
<ol start="4" style="list-style-type: decimal">
<li>Having made the progeny population, phenotype it also</li>
</ol>
<pre class="r"><code>f1_popn &lt;-AlphaSimR::setPheno(f1_popn, varE = 1)
pheno_f1 &lt;- f1_popn@pheno
pheno_f1 # since there are 200 F1&#39;s, it should be 200 phenotypic values.</code></pre>
<pre><code>            Trait1
  [1,]  0.90861748
  [2,]  0.88488008
  [3,] -0.10121392
  [4,] -1.92871357
  [5,]  0.88436764
  [6,]  0.02483377
  [7,]  1.67133236
  [8,]  2.23259174
  [9,]  2.58150701
 [10,] -1.28508385
 [11,] -1.24483195
 [12,] -0.68506445
 [13,] -2.58926422
 [14,] -0.01399514
 [15,]  0.01881472
 [16,]  2.62219784
 [17,]  1.50804630
 [18,]  2.21945843
 [19,]  0.71934206
 [20,] -1.40073105
 [21,]  2.34309432
 [22,] -1.56501426
 [23,]  0.06054963
 [24,]  0.50297116
 [25,] -0.19259329
 [26,]  1.82737400
 [27,]  0.24232497
 [28,]  0.44548812
 [29,] -0.54625762
 [30,] -2.44201536
 [31,] -1.53314738
 [32,]  0.79597527
 [33,] -1.17744678
 [34,] -1.95885455
 [35,] -0.90302042
 [36,]  0.71661593
 [37,] -1.30593013
 [38,]  0.67377388
 [39,]  0.13067185
 [40,]  0.61114893
 [41,]  0.09993337
 [42,]  2.05586105
 [43,]  0.13923656
 [44,] -0.60144522
 [45,] -0.73758065
 [46,] -0.50353192
 [47,]  0.24555989
 [48,] -1.36021033
 [49,]  0.59080149
 [50,]  0.47358792
 [51,]  2.27999996
 [52,]  0.37920012
 [53,] -1.75909889
 [54,]  0.61897662
 [55,]  1.64250045
 [56,] -0.89259198
 [57,]  1.37743796
 [58,]  1.13875002
 [59,] -1.11318806
 [60,] -0.62408043
 [61,]  1.00060640
 [62,] -2.01165725
 [63,] -1.03355646
 [64,] -0.02518026
 [65,]  0.13816497
 [66,] -3.64490371
 [67,]  1.51757427
 [68,] -2.22306560
 [69,]  1.02146933
 [70,] -0.77522857
 [71,]  1.70909991
 [72,]  0.64420883
 [73,] -1.09501054
 [74,] -0.39521189
 [75,]  0.80188708
 [76,]  0.32523086
 [77,]  0.72943311
 [78,] -0.10196847
 [79,]  0.80309893
 [80,]  0.32713354
 [81,]  0.23618131
 [82,]  1.58678729
 [83,] -1.34291693
 [84,] -1.00602311
 [85,]  0.15373348
 [86,] -1.13958236
 [87,]  1.30472389
 [88,]  0.24263441
 [89,] -1.23951320
 [90,] -1.29122338
 [91,] -0.65552670
 [92,] -1.13734757
 [93,]  2.48517509
 [94,] -1.48581955
 [95,]  0.06515068
 [96,]  0.14290268
 [97,]  1.43184440
 [98,]  1.16771438
 [99,]  0.39293319
[100,] -0.31028582
[101,] -2.25568302
[102,] -1.34094854
[103,] -0.29813694
[104,]  0.90091600
[105,] -1.98714302
[106,] -1.23246530
[107,] -2.46015679
[108,] -0.29583124
[109,]  1.37613256
[110,] -0.85847146
[111,] -2.67647126
[112,] -0.08172826
[113,]  1.17211906
[114,] -1.34145471
[115,]  0.19797904
[116,]  0.26202056
[117,]  0.42164068
[118,]  0.02754048
[119,]  0.35965987
[120,] -0.05280281
[121,]  0.58139993
[122,] -0.10628762
[123,] -1.07411699
[124,]  0.34045976
[125,] -0.77494562
[126,]  0.29417569
[127,]  1.28739783
[128,]  0.39711801
[129,] -0.57739492
[130,] -1.17058729
[131,] -0.87778592
[132,] -0.27036748
[133,]  0.08430463
[134,] -1.41107156
[135,] -0.25135322
[136,]  1.77348272
[137,]  0.20036958
[138,] -0.52789906
[139,] -0.42304287
[140,]  0.71236603
[141,] -1.50376570
[142,] -0.17394018
[143,] -0.84106007
[144,] -0.25911808
[145,]  1.38697518
[146,] -0.18533263
[147,]  1.03215590
[148,] -0.29036009
[149,] -2.27253405
[150,]  0.93824592
[151,]  1.54779949
[152,]  1.62645362
[153,]  0.72291520
[154,] -1.68300917
[155,]  0.51308677
[156,]  3.11253903
[157,] -0.70573531
[158,]  0.29804330
[159,]  0.24042815
[160,] -0.19863145
[161,]  0.88732393
[162,] -1.95145779
[163,] -0.34029605
[164,]  0.08077917
[165,] -2.02879591
[166,] -0.37817277
[167,]  1.16000886
[168,] -0.05005316
[169,]  0.71531951
[170,]  0.25741836
[171,]  2.35618262
[172,]  3.92148413
[173,] -2.38600283
[174,] -0.35984749
[175,]  0.07772749
[176,]  0.64695530
[177,] -0.44268267
[178,]  0.53818746
[179,] -0.14539727
[180,]  1.23178005
[181,]  3.66489321
[182,] -1.98267974
[183,]  1.73621631
[184,]  0.69891883
[185,]  0.81894041
[186,]  1.48988320
[187,] -0.30060137
[188,]  1.71040596
[189,] -2.46126495
[190,]  0.58024912
[191,] -0.01465478
[192,]  0.35914504
[193,]  1.93777206
[194,]  0.87519025
[195,] -0.42974853
[196,] -1.49586934
[197,] -0.98180904
[198,] -0.86975979
[199,]  2.66423298
[200,]  1.83517703</code></pre>
<ol start="5" style="list-style-type: decimal">
<li>Use each row of the <code>crossPlan</code> to find the two parents
and calculate their phenotypic mean</li>
</ol>
<pre class="r"><code># parent means and mid-parent is same here

#since in each row, I have two randomly selected parents, i can use apply function to calculate parent means for each row. 
parent_means &lt;- apply(X= crossPlan_2, MARGIN = 1, function(row){ #margin =1 means row-wise operation
  mean(pheno_founders[row])
} )
parent_means # print the vector of parent means which should be 200 considering 200 pairs of parents needed to cross for creating 200 hybrid&#39;s.</code></pre>
<pre><code>  [1] -0.097140131 -0.119151514 -0.130845625  0.305917475  0.475888206
  [6]  0.719333371  0.524745073  0.961834106 -0.289868896 -0.558595143
 [11] -1.751281859 -1.286716714  1.305900413  0.575536940 -0.277041469
 [16]  0.047134707  1.095680402  0.551873615 -1.039179999 -1.259754645
 [21]  1.329183576  0.159360942  1.519874017  0.682701057 -0.182504126
 [26]  0.487484469  0.450201514 -0.496893272  0.883645184 -0.696882146
 [31] -0.889782434  2.500613238  0.220549151 -0.110030083  1.110367794
 [36]  0.217938767 -1.786202162 -0.597446761  0.009931789 -0.303704807
 [41]  0.636459708  0.718981557  0.507914604  1.173442464  0.447292097
 [46] -0.696882146 -0.563394235 -1.098425448  0.906466189  1.504456145
 [51]  0.776040466 -1.513481979  0.131009787 -0.084764851  0.273004862
 [56] -0.906445045 -0.619768800 -0.258844170 -1.259226677  0.188569904
 [61]  1.087736483 -0.003324859  0.110023410  0.334331817 -1.277105852
 [66] -0.428821438  0.468817838 -0.321222708  0.418141907 -0.383018695
 [71]  1.881728990  0.299682133 -2.215960852 -0.325525285 -0.692863748
 [76]  0.640152585 -0.247003460  1.430915285  0.223068360 -1.172020841
 [81]  0.399246624 -0.548792632  0.054983926 -1.671371591  0.313923368
 [86]  0.415726532  0.547297691 -0.498857084 -0.247423629  1.991603535
 [91]  0.145030719  0.721137072  0.476400934 -0.505366945  0.176057601
 [96]  0.060139616  0.907659836  0.917327060 -0.267634459 -0.179378423
[101] -1.520363353 -1.282920464 -1.439231570 -0.163333875 -0.883800115
[106] -0.505117333 -0.862529548 -0.553118663 -0.422434798  2.086955617
[111]  0.350758666  0.759791813  0.717248672 -1.799916804  1.128299439
[116] -0.995296467 -0.046723648 -0.132120835  1.126833979  0.050311684
[121]  1.265091317 -1.513481979 -0.810282315  0.287302550  1.294992090
[126]  0.741741471  0.055905742  0.812151289 -0.483489706  1.542037359
[131]  0.934867219  0.211351187  0.495962467 -2.373271422 -0.428125763
[136] -0.246840145  1.327694895 -0.040438341  0.099801753  1.338459637
[141] -0.484677692  0.649830504 -0.145358804  1.215458224  0.898976448
[146] -0.225563851  0.718201460 -0.161635136 -1.979188310 -1.690376207
[151]  0.835317107  1.024225620  1.791180877 -0.618621426 -1.265435474
[156]  3.099879110 -0.114762291 -0.933420582  0.079778672  0.649639511
[161]  0.160175312 -0.478428596 -0.516308420 -0.350946461 -0.656167654
[166]  0.552193086  0.766417895  0.617613492 -0.707956915  0.503224133
[171]  1.370824747  2.167892663 -0.937318263 -0.968648264  0.187584642
[176] -0.720980482 -0.040166734  1.306338102  0.049245699 -0.811877129
[181]  1.065646667 -1.721172274  0.420023502  1.612362262 -0.753840724
[186] -0.039739108  0.681627803  1.523025255 -0.240174239 -0.338594555
[191]  1.002435037 -0.715312933  1.705067751  1.043482534 -0.945726967
[196]  0.759791813 -0.637654401  0.899945377 -0.281172171  1.742688208</code></pre>
<ol start="6" style="list-style-type: decimal">
<li>Make a scatterplot of the progeny phenotypes against the parent mean
phenotypes</li>
</ol>
<ul>
<li><p>Here, Each black dot represents a pair of parent and progeny
phenotype means.</p></li>
<li><p>The blue line is the regression line fitted to the data. It
represents the predicted values between mean phenotype of parents and
progeny phenotypes</p></li>
<li><p>The shaded region around the dark blue line is for condifence
interval which shows the range where the true regression line can fall
into.</p></li>
</ul>
<pre class="r"><code># Combine parent means and progeny phenotypes to make a dataset
regression_data &lt;- data.frame(ParentMean = parent_means, Trait1 = pheno_f1)
head(regression_data)</code></pre>
<pre><code>   ParentMean      Trait1
1 -0.09714013  0.90861748
2 -0.11915151  0.88488008
3 -0.13084563 -0.10121392
4  0.30591748 -1.92871357
5  0.47588821  0.88436764
6  0.71933337  0.02483377</code></pre>
<pre class="r"><code># Plot regression to the mean
library(ggplot2)
ggplot(regression_data, aes(x = ParentMean, y = Trait1)) +
  geom_point() +
  geom_smooth(method = &quot;lm&quot;, color = &quot;blue&quot;) +
  labs(title = &quot;Regression to the Mean _dg663&quot;, x = &quot;Mean Phenotype of Parents&quot;, y = &quot;Phenotype of Progeny&quot;) +
  theme_minimal()</code></pre>
<p><img src="figure/Lab3_ProgrammingAlphaSimR.Rmd/unnamed-chunk-2-1.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>#fit a linear model as observed in scatter plot above

lm &lt;- lm(Trait1 ~ParentMean, data = regression_data)
coef(lm) #second one here is the regression coefficient where first one is intercept or mean</code></pre>
<pre><code> (Intercept)   ParentMean 
-0.006261791  0.556468719 </code></pre>
<pre class="r"><code>coef(lm)[2]</code></pre>
<pre><code>ParentMean 
 0.5564687 </code></pre>
<pre class="r"><code>#0.4702 #when i calculated here</code></pre>
<blockquote>
<p>What is the regression coefficient?</p>
</blockquote>
<ul>
<li><p>Here we got the moderate regression value to the means meaning
“For every 1 unit increase in mean phenotype of parents, only 0.47 units
increase in progeny mean phenotype”. ie. 47% of variation in phenotype
of progeny is explained by the mean phenotypes of their parent and so
remaining 53% by other facturs like environment, non additive effects or
measurement errors…</p></li>
<li><p>BUT while knitting, the values are changing every time, so above
numbers in interpretation changes accordingly. I couldn’t get same
values even by setting seed. Please let me know if there is a way to do
that.</p></li>
</ul>
<blockquote>
<p>How much closer to the mean are offspring, on average, compared to
their parents? QUESTION CANCELED</p>
</blockquote>
<div id="homework-grading" class="section level3">
<h3>Homework grading</h3>
<p>I want to receive a zipped folder with your homework in it one week
from today, by midnight. If you use the <code>workflowr</code> package
to create the folder structure, that’s great. You should have created an
Rmarkdown script that will create an html file that I can look through.
So:<br />
1. 4 points for just sending in a zipped folder on time<br />
2. 1 points for each successful step from above, with partial
credit<br />
3. I will subtract 0.5 points for every day the homework is late</p>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span>
Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 4.4.1 (2024-06-14)
Platform: x86_64-apple-darwin20
Running under: macOS 15.3

Matrix products: default
BLAS:   /Library/Frameworks/R.framework/Versions/4.4-x86_64/Resources/lib/libRblas.0.dylib 
LAPACK: /Library/Frameworks/R.framework/Versions/4.4-x86_64/Resources/lib/libRlapack.dylib;  LAPACK version 3.12.0

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

time zone: America/New_York
tzcode source: internal

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
 [1] lubridate_1.9.3 forcats_1.0.0   stringr_1.5.1   dplyr_1.1.4    
 [5] purrr_1.0.2     readr_2.1.5     tidyr_1.3.1     tibble_3.2.1   
 [9] ggplot2_3.5.1   tidyverse_2.0.0 workflowr_1.7.1

loaded via a namespace (and not attached):
 [1] gtable_0.3.5      xfun_0.47         bslib_0.8.0       processx_3.8.4   
 [5] lattice_0.22-6    callr_3.7.6       tzdb_0.4.0        vctrs_0.6.5      
 [9] tools_4.4.1       Rdpack_2.6.2      ps_1.7.7          generics_0.1.3   
[13] parallel_4.4.1    fansi_1.0.6       highr_0.11        pkgconfig_2.0.3  
[17] Matrix_1.7-0      lifecycle_1.0.4   farver_2.1.2      compiler_4.4.1   
[21] git2r_0.33.0      munsell_0.5.1     getPass_0.2-4     httpuv_1.6.15    
[25] htmltools_0.5.8.1 sass_0.4.9        yaml_2.3.10       later_1.3.2      
[29] pillar_1.9.0      crayon_1.5.3      jquerylib_0.1.4   whisker_0.4.1    
[33] cachem_1.1.0      nlme_3.1-165      AlphaSimR_1.6.1   tidyselect_1.2.1 
[37] digest_0.6.37     stringi_1.8.4     labeling_0.4.3    splines_4.4.1    
[41] rprojroot_2.0.4   fastmap_1.2.0     grid_4.4.1        here_1.0.1       
[45] colorspace_2.1-1  cli_3.6.3         magrittr_2.0.3    utf8_1.2.4       
[49] withr_3.0.1       scales_1.3.0      promises_1.3.0    bit64_4.0.5      
[53] timechange_0.3.0  rmarkdown_2.28    httr_1.4.7        bit_4.0.5        
[57] hms_1.1.3         evaluate_0.24.0   knitr_1.48        rbibutils_2.3    
[61] mgcv_1.9-1        rlang_1.1.4       Rcpp_1.0.13       glue_1.7.0       
[65] rstudioapi_0.16.0 vroom_1.6.5       jsonlite_1.8.8    R6_2.5.1         
[69] fs_1.6.4         </code></pre>
</div>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
https://docs.mathjax.org/en/latest/web/configuration.html. This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>




</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
